File: surface/widgets/compose/components/widgets/domain/StateMachineDiagram.kt
// ============================================================
// Clef Surface Compose Widget — StateMachineDiagram
//
// Visual state-and-transition diagram rendered on a Canvas
// composable. States display as circles with labels and
// transitions as arrows between them. The current state is
// highlighted.
//
// Adapts the state-machine-diagram.widget spec: anatomy (root,
// stateList, stateItem, stateName, stateFlags, transitionList,
// transitionItem, transitionFrom, transitionArrow, transitionTo,
// transitionLabel, addStateButton, addTransitionButton), states,
// and connect attributes.
// ============================================================

package clef.surface.compose.components.widgets.domain

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

data class MachineState(
    val name: String,
    val initial: Boolean = false,
    val final: Boolean = false,
)

data class MachineTransition(
    val from: String,
    val to: String,
    val event: String,
)

// --------------- Component ---------------

/**
 * State machine diagram with state circles and transition arrows on Canvas.
 *
 * @param states List of states in the machine.
 * @param transitions List of transitions between states.
 * @param currentState Name of the currently active state.
 * @param canvasHeight Height of the drawing canvas.
 * @param modifier Modifier for the root layout.
 */
@Composable
fun StateMachineDiagram(
    states: List<MachineState>,
    transitions: List<MachineTransition>,
    currentState: String? = null,
    canvasHeight: Dp = 200.dp,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier.padding(8.dp)) {
        // Canvas diagram
        Canvas(modifier = Modifier.fillMaxWidth().height(canvasHeight)) {
            if (states.isEmpty()) return@Canvas

            val centerX = size.width / 2f
            val centerY = size.height / 2f
            val radius = minOf(centerX, centerY) * 0.65f
            val stateRadius = 28f

            // Position states in a circle
            val positions = states.mapIndexed { index, state ->
                val angle = (2 * Math.PI * index / states.size) - Math.PI / 2
                state.name to Offset(
                    centerX + radius * kotlin.math.cos(angle).toFloat(),
                    centerY + radius * kotlin.math.sin(angle).toFloat(),
                )
            }.toMap()

            // Draw transitions as arrows
            transitions.forEach { t ->
                val from = positions[t.from] ?: return@forEach
                val to = positions[t.to] ?: return@forEach

                // Shorten line to stop at circle edges
                val dx = to.x - from.x
                val dy = to.y - from.y
                val dist = kotlin.math.sqrt((dx * dx + dy * dy).toDouble()).toFloat()
                if (dist == 0f) return@forEach
                val nx = dx / dist
                val ny = dy / dist

                val start = Offset(from.x + nx * stateRadius, from.y + ny * stateRadius)
                val end = Offset(to.x - nx * stateRadius, to.y - ny * stateRadius)

                drawLine(
                    color = Color.Gray,
                    start = start,
                    end = end,
                    strokeWidth = 2f,
                )

                // Arrowhead
                val arrowLen = 10f
                val arrowAngle = Math.toRadians(25.0)
                val angle = kotlin.math.atan2(ny.toDouble(), nx.toDouble())
                drawLine(
                    color = Color.Gray,
                    start = end,
                    end = Offset(
                        end.x - arrowLen * kotlin.math.cos(angle - arrowAngle).toFloat(),
                        end.y - arrowLen * kotlin.math.sin(angle - arrowAngle).toFloat(),
                    ),
                    strokeWidth = 2f,
                )
                drawLine(
                    color = Color.Gray,
                    start = end,
                    end = Offset(
                        end.x - arrowLen * kotlin.math.cos(angle + arrowAngle).toFloat(),
                        end.y - arrowLen * kotlin.math.sin(angle + arrowAngle).toFloat(),
                    ),
                    strokeWidth = 2f,
                )
            }

            // Draw states as circles
            states.forEach { state ->
                val pos = positions[state.name] ?: return@forEach
                val isCurrent = state.name == currentState

                // Fill
                drawCircle(
                    color = if (isCurrent) Color(0xFF4CAF50) else Color(0xFFE0E0E0),
                    radius = stateRadius,
                    center = pos,
                )
                // Border
                drawCircle(
                    color = if (isCurrent) Color(0xFF2E7D32) else Color.Gray,
                    radius = stateRadius,
                    center = pos,
                    style = Stroke(width = if (state.final) 4f else 2f),
                )
                // Initial state marker (inner dot)
                if (state.initial) {
                    drawCircle(
                        color = Color.DarkGray,
                        radius = 4f,
                        center = Offset(pos.x - stateRadius - 8f, pos.y),
                    )
                    drawLine(
                        color = Color.DarkGray,
                        start = Offset(pos.x - stateRadius - 4f, pos.y),
                        end = Offset(pos.x - stateRadius, pos.y),
                        strokeWidth = 2f,
                    )
                }
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Transition list
        Text(
            text = "Transitions:",
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.Bold,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        LazyColumn {
            itemsIndexed(transitions) { _, t ->
                val fromIsCurrent = t.from == currentState
                Row(modifier = Modifier.padding(start = 8.dp, top = 2.dp)) {
                    Text(
                        text = "(${t.from})",
                        color = if (fromIsCurrent) Color(0xFF4CAF50) else MaterialTheme.colorScheme.onSurface,
                    )
                    Text(
                        text = " \u2014",
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Text(
                        text = t.event,
                        color = MaterialTheme.colorScheme.tertiary,
                    )
                    Text(
                        text = "\u2192 ",
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Text(text = "(${t.to})")
                }
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Legend
        Row {
            Text(
                text = "\u25B8 initial   \u25A0 final   ",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Text(
                text = "\u25CF current",
                style = MaterialTheme.typography.labelSmall,
                color = Color(0xFF4CAF50),
            )
        }
    }
}

File: surface/widgets/swiftui/components/widgets/domain/StateMachineDiagram.swift
// ============================================================
// Clef Surface SwiftUI Widget — StateMachineDiagram
//
// Visual state-and-transition diagram rendered on a Canvas.
// States display as circles with labels and transitions as
// arrows between them. The current state is highlighted.
// ============================================================

import SwiftUI

struct MachineState: Identifiable {
    var id: String { name }
    let name: String
    var initial: Bool = false
    var isFinal: Bool = false
}

struct MachineTransition: Identifiable {
    var id: String { "\(from)-\(event)-\(to)" }
    let from: String
    let to: String
    let event: String
}

struct StateMachineDiagramView: View {
    var states: [MachineState]
    var transitions: [MachineTransition]
    var currentState: String? = nil
    var canvasHeight: CGFloat = 200

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // Canvas diagram
            Canvas { context, size in
                guard !states.isEmpty else { return }

                let centerX = size.width / 2
                let centerY = size.height / 2
                let radius = min(centerX, centerY) * 0.65
                let stateRadius: CGFloat = 28

                // Position states in a circle
                let positions: [String: CGPoint] = Dictionary(uniqueKeysWithValues:
                    states.enumerated().map { index, state in
                        let angle = (2 * .pi * CGFloat(index) / CGFloat(states.count)) - .pi / 2
                        return (state.name, CGPoint(
                            x: centerX + radius * cos(angle),
                            y: centerY + radius * sin(angle)
                        ))
                    }
                )

                // Draw transitions as arrows
                for t in transitions {
                    guard let from = positions[t.from], let to = positions[t.to] else { continue }

                    let dx = to.x - from.x
                    let dy = to.y - from.y
                    let dist = sqrt(dx * dx + dy * dy)
                    guard dist > 0 else { continue }
                    let nx = dx / dist
                    let ny = dy / dist

                    let start = CGPoint(x: from.x + nx * stateRadius, y: from.y + ny * stateRadius)
                    let end = CGPoint(x: to.x - nx * stateRadius, y: to.y - ny * stateRadius)

                    var linePath = Path()
                    linePath.move(to: start)
                    linePath.addLine(to: end)
                    context.stroke(linePath, with: .color(.gray), style: StrokeStyle(lineWidth: 2))

                    // Arrowhead
                    let arrowLen: CGFloat = 10
                    let arrowAngle: CGFloat = .pi / 7.2
                    let angle = atan2(ny, nx)

                    var arrow1 = Path()
                    arrow1.move(to: end)
                    arrow1.addLine(to: CGPoint(
                        x: end.x - arrowLen * cos(angle - arrowAngle),
                        y: end.y - arrowLen * sin(angle - arrowAngle)
                    ))
                    context.stroke(arrow1, with: .color(.gray), style: StrokeStyle(lineWidth: 2))

                    var arrow2 = Path()
                    arrow2.move(to: end)
                    arrow2.addLine(to: CGPoint(
                        x: end.x - arrowLen * cos(angle + arrowAngle),
                        y: end.y - arrowLen * sin(angle + arrowAngle)
                    ))
                    context.stroke(arrow2, with: .color(.gray), style: StrokeStyle(lineWidth: 2))
                }

                // Draw states as circles
                for state in states {
                    guard let pos = positions[state.name] else { continue }
                    let isCurrent = state.name == currentState

                    let rect = CGRect(x: pos.x - stateRadius, y: pos.y - stateRadius,
                                      width: stateRadius * 2, height: stateRadius * 2)

                    // Fill
                    context.fill(
                        Circle().path(in: rect),
                        with: .color(isCurrent ? Color.green : Color(.systemGray5))
                    )
                    // Border
                    context.stroke(
                        Circle().path(in: rect),
                        with: .color(isCurrent ? Color(red: 0.18, green: 0.49, blue: 0.20) : .gray),
                        style: StrokeStyle(lineWidth: state.isFinal ? 4 : 2)
                    )

                    // Initial state marker
                    if state.initial {
                        let dotCenter = CGPoint(x: pos.x - stateRadius - 8, y: pos.y)
                        let dotRect = CGRect(x: dotCenter.x - 4, y: dotCenter.y - 4, width: 8, height: 8)
                        context.fill(Circle().path(in: dotRect), with: .color(.gray))

                        var connPath = Path()
                        connPath.move(to: CGPoint(x: pos.x - stateRadius - 4, y: pos.y))
                        connPath.addLine(to: CGPoint(x: pos.x - stateRadius, y: pos.y))
                        context.stroke(connPath, with: .color(.gray), style: StrokeStyle(lineWidth: 2))
                    }
                }
            }
            .frame(height: canvasHeight)

            // Transition list
            Text("Transitions:")
                .font(.caption)
                .fontWeight(.bold)
                .foregroundColor(.secondary)

            ForEach(transitions) { t in
                let fromIsCurrent = t.from == currentState
                HStack(spacing: 4) {
                    Text("(\(t.from))")
                        .foregroundColor(fromIsCurrent ? .green : .primary)
                    Text("\u{2014}")
                        .foregroundColor(.secondary)
                    Text(t.event)
                        .foregroundColor(.purple)
                    Text("\u{2192}")
                        .foregroundColor(.secondary)
                    Text("(\(t.to))")
                }
                .font(.caption)
                .padding(.leading, 8)
            }

            // Legend
            HStack(spacing: 4) {
                Text("\u{25B8} initial   \u{25A0} final   ")
                    .font(.caption2)
                    .foregroundColor(.secondary)
                Text("\u{25CF} current")
                    .font(.caption2)
                    .foregroundColor(.green)
            }
        }
        .padding(8)
    }
}
